zkp/golang/lib/smt.go

```diff
@@ -100,56 +119,127 @@ func (mt *sparseMerkleTree) GetNode(key smt.NodeIndex) (smt.Node, error) {
 	return node, nil
 }
 
-// addLeaf adds a new LeafNode to the MerkleTree. It traverses the tree from
-// the node at "rootKey", following the path determined by the "path" parameter
-func (mt *sparseMerkleTree) addLeaf(newLeaf smt.Node, rootKey smt.NodeIndex, lvl int, path []bool) (smt.NodeIndex, error) {
-	var err error
-	var nextKey smt.NodeIndex
-	if lvl > mt.maxLevels-1 {
+// GenerateProof generates the proof of existence (or non-existence) of a leaf node
+// for a Merkle Tree given the root. It uses the node's index to represent the node.
+// If the rootKey is nil, the current merkletree root is used
+func (mt *sparseMerkleTree) GenerateProof(k *big.Int, rootKey smt.NodeIndex) (*smt.Proof, *big.Int, error) {
```

Comment:
So this method will generate either a proof of existence or non-existence. Would it make sense to specify which I expect to generate?

Reply:
that makes sense. we can have two convenient method in addition to this, `GenerateInclusionProof()` and `GenerateNonInclusionProof()`, which does the check after getting the proof and returns an error if not matching the expected proof

Reply:
Yea, I think that would be great so you can catch unexpected problems.